### PR TITLE
[PfemFluid] Multiple refining boxes in 2D and 3D

### PR DESCRIPTION
**Pull request by AFranci**
This PR introduces the possibility of defining multiple refining boxes. The parameters to set are the limit coordinates of the box, the initial and final times, and the number of elements in the transition from the fine mesh to the coarse one. The mesh size of the different zones (especially the transition zones) is set inside mesher_utilities.cpp (before it was duplicated and not optimized). This value is used by the three main remeshing processes (alpah shape, generatation and removal of nodes).

### CODE REVIEW

File: applications/DelaunayMeshingApplication/custom_utilities/mesher_utilities.cpp [hunk 1]
```diff
@@ -2150,6 +2150,534 @@ namespace Kratos
...
+          double distanceToBox = 0;
+          double coefficient = 0;
+
+          if (NodeCoordinates[0] < RefiningBoxMinimumPoint[0] && NodeCoordinates[0] > minExternalPoint[0] && NodeCoordinates[1] > minExternalPoint[1] && NodeCoordinates[1] < maxExternalPoint[1])
+          {
+            if (NodeCoordinates[1] < RefiningBoxMinimumPoint[1])
+            {
+              const double distanceToBoxX = NodeCoordinates[0] - RefiningBoxMinimumPoint[0];
+              const double distanceToBoxY = NodeCoordinates[1] - RefiningBoxMinimumPoint[1];
+              const double radius = pow(distanceToBoxX, 2) + pow(distanceToBoxY, 2);
```
Comment (AlejandroCornejo) [inline applications/DelaunayMeshingApplication/custom_utilities/mesher_utilities.cpp]: we usually use std::pow and std::sqrt

File: applications/DelaunayMeshingApplication/custom_utilities/mesher_utilities.cpp [hunk 2]
```diff
@@ -2150,6 +2150,534 @@ namespace Kratos
...
+                distanceToBox = transitionDistance;
+              }
+              counter++;
+            }
+            else if (NodeCoordinates[1] > RefiningBoxMaximumPoint[1])
+            {
+              const double distanceToBoxX = NodeCoordinates[0] - RefiningBoxMinimumPoint[0];
+              const double distanceToBoxY = NodeCoordinates[1] - RefiningBoxMaximumPoint[1];
+              const double radius = pow(distanceToBoxX, 2) + pow(distanceToBoxY, 2);
+              distanceToBox = sqrt(radius);
```
Comment (AlejandroCornejo) [inline applications/DelaunayMeshingApplication/custom_utilities/mesher_utilities.cpp]: std::

File: applications/DelaunayMeshingApplication/custom_utilities/mesher_utilities.cpp [hunk 3]
```diff
@@ -2150,6 +2150,534 @@ namespace Kratos
...
+              {
+                distanceToBox = transitionDistance;
+              }
+              counter++;
+            }
+            else if (NodeCoordinates[1] > RefiningBoxMaximumPoint[1])
+            {
+              const double distanceToBoxX = NodeCoordinates[0] - RefiningBoxMinimumPoint[0];
+              const double distanceToBoxY = NodeCoordinates[1] - RefiningBoxMaximumPoint[1];
+              const double radius = pow(distanceToBoxX, 2) + pow(distanceToBoxY, 2);
```
Comment (AlejandroCornejo) [inline applications/DelaunayMeshingApplication/custom_utilities/mesher_utilities.cpp]: std

File: applications/DelaunayMeshingApplication/custom_utilities/mesher_utilities.cpp [hunk 4]
```diff
@@ -2150,6 +2150,534 @@ namespace Kratos
...
+              }
+            }
+            else if (NodeCoordinates[1] > RefiningBoxMaximumPoint[1])
+            {
+              if (NodeCoordinates[2] < RefiningBoxMinimumPoint[2])
+              {
+                const double distanceToBoxX = NodeCoordinates[0] - RefiningBoxMinimumPoint[0];
+                const double distanceToBoxY = NodeCoordinates[1] - RefiningBoxMaximumPoint[1];
+                const double distanceToBoxZ = NodeCoordinates[2] - RefiningBoxMinimumPoint[2];
+                const double radius = pow(distanceToBoxX, 2) + pow(distanceToBoxY, 2) + pow(distanceToBoxZ, 2);
```
Comment (AlejandroCornejo) [inline applications/DelaunayMeshingApplication/custom_utilities/mesher_utilities.cpp]: idem

File: applications/DelaunayMeshingApplication/custom_utilities/mesher_utilities.cpp [hunk 5]
```diff
@@ -2150,6 +2150,534 @@ namespace Kratos
...
+                }
+                counter++;
+              }
+              else if (NodeCoordinates[2] > RefiningBoxMaximumPoint[2])
+              {
+                const double distanceToBoxX = NodeCoordinates[0] - RefiningBoxMinimumPoint[0];
+                const double distanceToBoxY = NodeCoordinates[1] - RefiningBoxMaximumPoint[1];
+                const double distanceToBoxZ = NodeCoordinates[2] - RefiningBoxMaximumPoint[2];
+                const double radius = pow(distanceToBoxX, 2) + pow(distanceToBoxY, 2) + pow(distanceToBoxZ, 2);
+                distanceToBox = sqrt(radius);
```
Comment (AlejandroCornejo) [inline applications/DelaunayMeshingApplication/custom_utilities/mesher_utilities.cpp]: idfem

File: applications/DelaunayMeshingApplication/custom_utilities/mesher_utilities.hpp [hunk 6]
```diff
@@ -917,63 +911,68 @@ namespace Kratos
...
                   };
 
-                  void SetUseRefiningBox(bool rUseRefiningBox)
-                  {
-                        UseRefiningBox = rUseRefiningBox;
-                  };
+                  void InitializeRefiningBoxParameters(unsigned int size)
+                  {
+                        UseRefiningBox.resize(size);
+                        RefiningBoxMinimumPoint.resize(size);
```
Comment (AlejandroCornejo) [inline applications/DelaunayMeshingApplication/custom_utilities/mesher_utilities.hpp:917]: `resize(size, false)` is faster

File: applications/PfemFluidDynamicsApplication/custom_processes/generate_new_nodes_before_meshing_process.hpp [hunk 7]
```diff
@@ -230,60 +229,40 @@ namespace Kratos
...
-				NewPositions.resize(ElementsToRefine);
-				BiggestVolumes.resize(ElementsToRefine, false);
-				NodesIDToInterpolate.resize(ElementsToRefine);
-				// NewDofs.resize(ElementsToRefine, false);
-
-				for (int nn = 0; nn < ElementsToRefine; nn++)
-				{
-					BiggestVolumes[nn] = -1.0;
-				}
+				NewPositions.resize(0);
```
Comment (AlejandroCornejo) [inline applications/PfemFluidDynamicsApplication/custom_processes/generate_new_nodes_before_meshing_process.hpp:237]: ,false

File: applications/PfemFluidDynamicsApplication/custom_processes/generate_new_nodes_before_meshing_process.hpp [hunk 8]
```diff
@@ -1297,130 +1272,26 @@ namespace Kratos
...
 			{
-				if (freesurfaceNodes > 0)
-				{
-					penalization = 1.2; // to avoid to gain too much volume during remeshing step
-				}
-
-				if (rigidNodes > 0 && penalizationRigid == true)
+				const double rigidWallMeshSize = rigidNodeLocalMeshSize / rigidNodeMeshCounter;
+				const double ratio = rigidWallMeshSize / meanMeshSize;
+				const double tolerance = 1.8;
```
Comment (AlejandroCornejo) [inline applications/PfemFluidDynamicsApplication/custom_processes/generate_new_nodes_before_meshing_process.hpp:1281]: interesting tolerance

File: applications/PfemFluidDynamicsApplication/custom_processes/generate_new_nodes_before_meshing_process.hpp [hunk 9]
```diff
@@ -1297,130 +1272,26 @@ namespace Kratos
...
+			if (freesurfaceNodes > 0)
+			{
+				penalization = 1.2; // to avoid to gain too much volume during remeshing step
+			}
 			const double safetyCoefficient2D = 1.5;
-
-			double ElementalVolume = Element.Area();
-
 			array_1d<double, 3> Edges(3, 0.0);
 			array_1d<unsigned int, 3> FirstEdgeNode(3, 0);
```
Comment (AlejandroCornejo) [inline applications/PfemFluidDynamicsApplication/custom_processes/generate_new_nodes_before_meshing_process.hpp:1296]: const?

File: applications/PfemFluidDynamicsApplication/custom_processes/generate_new_nodes_before_meshing_process.hpp [hunk 10]
```diff
@@ -1587,166 +1444,28 @@ namespace Kratos
...
-					inletNodes++;
-				}
+			}
 
-				if (refiningBox == true)
+			if (rigidNodeMeshCounter > 0)
+			{
+				const double rigidWallMeshSize = rigidNodeLocalMeshSize / rigidNodeMeshCounter;
+				const double ratio = rigidWallMeshSize / meanMeshSize;
+				const double tolerance = 1.8;
```
Comment (AlejandroCornejo) [inline applications/PfemFluidDynamicsApplication/custom_processes/generate_new_nodes_before_meshing_process.hpp:1453]: if repeatred you could define this tolerance in the .h file as a const expr

File: applications/PfemFluidDynamicsApplication/custom_processes/generate_new_nodes_before_meshing_process.hpp [hunk 11]
```diff
@@ -1587,166 +1444,28 @@ namespace Kratos
...
-					penalization = 1.2; // to avoid to gain too much volume during remeshing step
-				}
 
-				if (rigidNodes > 0 && penalizationRigid == true)
-				{
-					penalization = 1.15;
-				}
+			if (freesurfaceNodes > 0)
+			{
+				penalization = 1.2; // to avoid to gain too much volume during remeshing step
```
Comment (AlejandroCornejo) [inline applications/PfemFluidDynamicsApplication/custom_processes/generate_new_nodes_before_meshing_process.hpp:1465]: alquimist :D

File: applications/PfemFluidDynamicsApplication/custom_processes/remove_mesh_nodes_for_fluids_process.hpp [hunk 12]
```diff
@@ -126,35 +126,35 @@ namespace Kratos
 			{
 				if (mEchoLevel > 1)
 					std::cout << " REMOVE_NODES is TRUE " << std::endl;
-				// bool any_node_removed_on_error = false;
+				// bool some_node_is_removed_due_to_error = false;
 				// ////////////////////////////////////////////////////////////
 				// if (mrRemesh.Refine->RemovingOptions.Is(MesherUtilities::REMOVE_NODES_ON_ERROR))
```
Comment (AlejandroCornejo) [inline applications/PfemFluidDynamicsApplication/custom_processes/remove_mesh_nodes_for_fluids_process.hpp]: remove unused code?

File: applications/PfemFluidDynamicsApplication/custom_processes/remove_mesh_nodes_for_fluids_process.hpp [hunk 13]
```diff
@@ -407,13 +403,11 @@ namespace Kratos
 				// coordinates
 				for (unsigned int i = 0; i < ie->GetGeometry().size(); i++)
 				{
-
 					if ((ie->GetGeometry()[i].Is(RIGID) && ie->GetGeometry()[i].IsNot(INLET)) || ie->GetGeometry()[i].Is(SOLID))
```
Comment (AlejandroCornejo) [inline applications/PfemFluidDynamicsApplication/custom_processes/remove_mesh_nodes_for_fluids_process.hpp:334]: you can declar a `const auto& r_geometry = GetGeometry()` and the reuse the `r_geometry`

File: applications/PfemFluidDynamicsApplication/custom_processes/select_mesh_elements_for_fluids_process.hpp [hunk 14]
```diff
@@ -403,14 +402,14 @@ namespace Kratos
...
-                                                            (sqrt(pow(nodesVelocities[0][0], 2) + pow(nodesVelocities[0][1], 2)) *
-                                                             sqrt(pow(nodesVelocities[1][0], 2) + pow(nodesVelocities[1][1], 2)));
+                                                                  (sqrt(pow(nodesVelocities[0][0], 2) + pow(nodesVelocities[0][1], 2)) *
+                                                                   sqrt(pow(nodesVelocities[1][0], 2) + pow(nodesVelocities[1][1], 2)));
                                         const double cosAngle02 = (nodesVelocities[0][0] * nodesVelocities[2][0] + nodesVelocities[0][1] * nodesVelocities[2][1]) /
-                                                            (sqrt(pow(nodesVelocities[0][0], 2) + pow(nodesVelocities[0][1], 2)) *
-                                                             sqrt(pow(nodesVelocities[2][0], 2) + pow(nodesVelocities[2][1], 2)));
+                                                                  (sqrt(pow(nodesVelocities[0][0], 2) + pow(nodesVelocities[0][1], 2)) *
```
Comment (AlejandroCornejo) [inline applications/PfemFluidDynamicsApplication/custom_processes/select_mesh_elements_for_fluids_process.hpp]: std

File: applications/PfemFluidDynamicsApplication/custom_processes/select_mesh_elements_for_fluids_process.hpp [hunk 15]
```diff
@@ -872,7 +756,7 @@ namespace Kratos
             {
                 if (numfreesurf < nds && numisolated == 0)
                 {
-                    Alpha *= 1.275;
+                    Alpha *= 1.2;
```
Comment (AlejandroCornejo) [inline applications/PfemFluidDynamicsApplication/custom_processes/select_mesh_elements_for_fluids_process.hpp:759]: imagen

**Review by AlejandroCornejo — CHANGES_REQUESTED** [review]

**Review by rlrangel — APPROVED** [review]
Ok.

**Review by AlejandroCornejo — APPROVED** [review]
ok!!!

**Pull request merged**